internal/transport/http2_client.go

```diff
@@ -1254,14 +1256,132 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+			}
+			// TODO: do we want to propagate the whole content-type in the metadata,
+			// or come up with a way to just propagate the content-subtype if it was set?
+			// ie {"content-type": "application/grpc+proto"} or {"content-subtype": "proto"}
+			// in the metadata?
+			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
+			isGRPC = true
+		case "grpc-encoding":
+			s.recvCompress = hf.Value
+		case "grpc-status":
+			code, err := strconv.Atoi(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status: %v", err))
+				break
+			}
+			rawStatusCode = &code
+		case "grpc-message":
+			grpcMessage = decodeGrpcMessage(hf.Value)
+		case "grpc-status-details-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				break
+			}
+			st := &spb.Status{}
+			if err := proto.Unmarshal(v, st); err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				break
+			}
+			statusGen = status.FromProto(st)
+		case ":status":
+			code, err := strconv.Atoi(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed http-status: %v", err))
+				break
+			}
+			httpStatus = &code
+		case "grpc-tags-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-tags-bin: %v", err))
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], string(v))
+		case "grpc-trace-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-trace-bin: %v", err))
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], string(v))
+		default:
+			if isReservedHeader(hf.Name) && !isWhitelistedHeader(hf.Name) {
+				break
+			}
+			v, err := decodeMetadataHeader(hf.Name, hf.Value)
+			if err != nil {
+				if logger.V(logLevel) {
+					logger.Errorf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
+				}
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], v)
+		}
+
+		// Account for the fact that we can set isGRPC at any time - it's possible
+		// that we'll set isGRPC AFTER an error has occurred.
+		if isGRPC && grpcStatus != nil {
+			t.closeStream(s, grpcStatus.Err(), true, http2.ErrCodeProtocol, grpcStatus, nil, endStream)
+			return
+		}
+	}
+
+	if isGRPC {
+		// gRPC status doesn't exist.
+		// Set rawStatusCode to be unknown so that, if the stream has ended this
+		// Unknown status will be propagated to the user.
+		// Otherwise, it will be ignored. In which case, status from
+		// a later trailer, that has StreamEnded flag set, is propagated.
+		if rawStatusCode == nil && statusGen == nil {
+			code := int(codes.Unknown)
+			rawStatusCode = &code
```

Comment:
How about:

```
rawStatusCode := codes.Unknown
....
case "grpc-status":
	code, err := strconv.ParseUint(hf.Value, 0, 32)
	if err != nil {
		grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status: %v", err))
		break
	}
	rawStatusCode = codes.Code(uint32(code))
...
```

and remove this code here?  This whole `if isGRPC` part can be removed, leaving the `else` (`if !isGRPC`) by itself.